### ISSUE TITLE
fips: integration testing errors. can no longer prevent auto-attach on PRO images with cloud-config

### ISSUE THREAD

**Issue by blackboxsw** [canonical/ubuntu-advantage-client#1297]
Because auto-attach now runs before cloud-config userdata is processed, we can't prevent auto-attach from running on PRO images during launch. As a result, we need either explicit test setup for PRO images to avoid installing ubuntu-advantage-pro deb, or our PRO tests should still expect that an auto-attach has occurred and cleanup, re-attach those test instances.

failures we see are tracebacks in the test logs due to auto-attach running:
error message:
   ```bash
# features/steps/steps.py:227
    And I verify that running `grep Traceback /var/log/ubuntu-advantage.log` `with sudo` exits `1`                                                # features/steps/steps.py:227
      Traceback (most recent call last):
        File "/jenkins/servers/server/workspace/uaclient-vm/UBUNTUVMRELEASE/18.04/ua-client/.tox/behave-vm-18.04/lib/python3.6/site-packages/behave/model.py", line 1329, in run
          match.run(runner.context)
        File "/jenkins/servers/server/workspace/uaclient-vm/UBUNTUVMRELEASE/18.04/ua-client/.tox/behave-vm-18.04/lib/python3.6/site-packages/behave/matchers.py", line 98, in run
          self.func(context, *args, **kwargs)
        File "features/steps/steps.py", line 234, in then_i_verify_that_running_cmd_with_spec_exits_with_codes
          assert str(context.process.returncode) in expected_codes
      AssertionError
```

jenkins job log: https://jenkins.ubuntu.com/server/view/ua/job/uaclient-vm/28/UBUNTUVMRELEASE=18.04/console

**Comment by lucasmoura** [comment]
I am currently working on ensuring that pro deb are not installed on non-pro machines. However, when looking at the pro machine logs to confirm if auto-attach is being run, I get this:

```
2021-01-07 15:45:09,224 - cli.py:(986) [DEBUG]: Executed with sys.argv: ['/usr/bin/ua', 'auto-attach']
2021-01-07 15:45:09,224 - util.py:(581) [DEBUG]: Writing file: /var/lib/ubuntu-advantage/private/lock
2021-01-07 15:45:09,224 - cli.py:(691) [DEBUG]: Skipping auto-attach. Config disable_auto_attach is set.
2021-01-07 15:45:09,224 - util.py:(591) [DEBUG]: Removing file: /var/lib/ubuntu-advantage/private/lock
2021-01-07 15:45:11,988 - cli.py:(986) [DEBUG]: Executed with sys.argv: ['/usr/bin/ua', 'auto-attach']
2021-01-07 15:45:11,989 - util.py:(581) [DEBUG]: Writing file: /var/lib/ubuntu-advantage/private/lock
2021-01-07 15:45:11,992 - cli.py:(691) [DEBUG]: Skipping auto-attach. Config disable_auto_attach is set.
2021-01-07 15:45:11,992 - util.py:(591) [DEBUG]: Removing file: /var/lib/ubuntu-advantage/private/lock
2021-01-07 15:45:57,373 - cli.py:(986) [DEBUG]: Executed with sys.argv: ['/usr/bin/ua', 'auto-attach']
2021-01-07 15:45:57,373 - util.py:(581) [DEBUG]: Writing file: /var/lib/ubuntu-advantage/private/lock
2021-01-07 15:45:57,806 - util.py:(382) [DEBUG]: Reading file: /sys/class/dmi/id/chassis_asset_tag
2021-01-07 15:45:57,806 - util.py:(382) [DEBUG]: Reading file: /etc/os-release
```

The `disable_auto_attach` is being set by the userdata we provide to the machine. However, since auto-attach should now run before cloud-init process the userdata, this should not happen, right ?

PS: This log was generated in Azure PRO bionic machine

**Comment by lucasmoura** [comment]
I have an hypothesis of why this is happening on the PRO images. We create a PRO instance, we pass the following userdata:
```
USERDATA_BLOCK_AUTO_ATTACH = """\
#cloud-config
write_files:
  # TODO(drop path: /usr/bin/ua when 25.0 is in Ubuntu PRO images)
  - path: /usr/bin/ua
    content: |
        #!/bin/bash
        DATE=`date -u`
        echo "$DATE: exec ua $@" >> /root/ua-runs
    permissions: '0755'
  - path: /etc/ubuntu-advantage/uaclient.conf
    content: |
      features:
         disable_auto_attach: true
    append: true
"""
```

We can see that the `disable_auto_attach` option is there. Since we create a base image with the newest `ubuntu-advantage-client` installed in it and snapshot it, that configuration will persist on that snapshot. This means that when we launch the actual instance that we want to test, that configuration is already there and we will not run `auto-attach` on boot

**Comment by blackboxsw** [comment]
In building our PRO sapshots, I think we are still exposed to the fact that those pro images auto-attach on initial launch, even before we install the ubuntu-advantage-tools pkg which can handle those overrides. Thus, I think we need to provide something that can make sure that an auto-attach doesn't happen on even our build images.

Something like the following can ensure our build images don't get muddied by an auto-attach:

```bash
#cloud-config
bootcmd:
 - cp /usr/bin/ua /usr/bin/ua.orig 
 - echo "#!/bin/sh\necho NOOP\n" > /usr/bin/ua
 - chmod 755 /usr/bin/ua
```

The above would only be required for our base image builds (not our test instances). `/usr/bin/ua` will get overwritten in our base image when we install the version of ubuntu-advantage-tools under test. So we don't have to worry about cleaning up that piece

**Comment by blackboxsw** [comment]
fixed in 10bed404
